src/CodeGenerator/ClassBuilder.cs

```diff
@@ -0,0 +1,301 @@
+        private const char EndDelimiter = '$';
+
+        private static readonly Dictionary<ClassAccessibilityLevel, string> ClassAccessibilityLevelMap = new()
+        {
+            { ClassAccessibilityLevel.Internal, "internal" },
+            { ClassAccessibilityLevel.Public, "public" },
+            { ClassAccessibilityLevel.Private, "private" },
+            { ClassAccessibilityLevel.PrivateProtected, "private protected" },
+            { ClassAccessibilityLevel.Protected, "protected" },
+            { ClassAccessibilityLevel.ProtectedInternal, "protected internal" }
+        };
+
+        private static readonly string[] AllowedModifiers =
+        {
+            "abstract",
+            "sealed",
+            "static"
+        };
+
+        private static readonly Regex LegalTypeNameCharacters = new("^[a-zA-Z0-9_]+$");
+        private static readonly Regex LegalTypeNameLeadingCharacters = new("^[a-zA-Z_]+$");
+
+        private string _namespace;
+        private string _className;
+        private ClassAccessibilityLevel? _accessibilityLevel;
+        private string _baseClass;
+
+        private readonly List<string> _modifiers = new();
+        private readonly List<string> _usingDirectives = new();
+        private readonly List<Method> _methods = new();
+        private readonly List<PropertyDefinition> _properties = new();
+        private readonly List<string> _attributes = new();
+        private readonly List<string> _constructors = new();
+        private readonly List<FieldDefinition> _fields = new();
+        private readonly List<string> _implementedInterfaces = new();
+
+        public ClassBuilder WithNamespace(string value)
+        {
+            EnsureValidNamespace(value);
+            _namespace = value;
+            return this;
+        }
+
+        public ClassBuilder UsingNamespace(string value)
+        {
+            EnsureValidNamespace(value);
+            if (_usingDirectives.Contains(value))
+            {
+                throw new InvalidOperationException("Namespace has already been added");
+            }
+            _usingDirectives.Add(value);
+            return this;
+        }
+
+        private static void EnsureValidNamespace(string value)
+        {
+            if (string.IsNullOrEmpty(value))
+            {
+                throw new ArgumentException("Invalid namespace");
+            }
+
+            var parts = value.Split(".");
+            foreach (var part in parts)
+            {
+                if (!LegalTypeNameCharacters.IsMatch(part)
+                    || !LegalTypeNameLeadingCharacters.IsMatch(part[..1]))
+                {
+                    throw new ArgumentException("Invalid namespace");
+                }
+            }
+        }
+
+        public ClassBuilder WithAccessibilityLevel(ClassAccessibilityLevel expected)
+        {
+            _accessibilityLevel = expected;
+            return this;
+        }
+
+        public ClassBuilder WithName(string value)
```

Comment:
For instances where `string` is validated using `LegalTypeNameCharacters`, it might be better to introduce a type instead of `string` which encapsulates the validation rules. Would reduce overall cyclomatic complexity in this class and produce a more meaningful contract